File: aaa-practic/v6/infrastructure/mem_db.rb
require_relative '../interfaces/todo_repository'
require_relative '../entities/todo'

class MemDb
  include TodoRepository
  
  def initialize
    @todos = []
    @next_id_counter = 1
  end
  
  def next_id
    id = @next_id_counter
    @next_id_counter += 1
    id
  end
  
  def add(todo)
    @todos << todo
  end
  
  def list_all
    @todos.dup
  end
  
  def find_by_id(todo_id)
    todo = @todos.find { |t| t.id == todo_id }
    raise StandardError, "Todo not found" unless todo
    todo
  end
  
  def update(updated_todo)
    index = @todos.find_index { |t| t.id == updated_todo.id }
    raise StandardError, "Todo not found" unless index
    @todos[index] = updated_todo
  end
  
  def delete(todo_id)
    index = @todos.find_index { |t| t.id == todo_id }
    raise StandardError, "Todo not found" unless index
    @todos.delete_at(index)
  end
  
  def count
    @todos.length
  end
end

File: aaa-practic/v5/infra/mem_db.py
from i.i_todo import ITd

class MemDb(ITd):
    def __init__(self):
        self.todos = []
        self._next_id = 1

    def next_id(self):
        nid = self._next_id
        self._next_id += 1
        return nid

    def add(self, todo):
        self.todos.append(todo)

    def list_all(self):
        return self.todos

    def find_by_id(self, todo_id):
        for todo in self.todos:
            if todo.id == todo_id:
                return todo
        raise ValueError("Todo not found")

    def update(self, updated_todo):
        for i, todo in enumerate(self.todos):
            if todo.id == updated_todo.id:
                self.todos[i] = updated_todo
                return
        raise ValueError("Todo not found")

    def delete(self, todo_id):
        for i, todo in enumerate(self.todos):
            if todo.id == todo_id:
                del self.todos[i]
                return
        raise ValueError("Todo not found")

    def count(self):
        return len(self.todos)